src/main.cpp

```diff
@@ -5483,22 +5538,25 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
     {
         std::vector<CInv> vInv;
         vRecv >> vInv;
-        // BU check size == 0 to be intolerant of an empty and useless request
-        if ((vInv.size() > MAX_INV_SZ)||(vInv.size() == 0))
+
+        // Message Consistency Checking
+        //   Check size == 0 to be intolerant of an empty and useless request.
+        //   Validate that INVs are a valid type and not null.
+        if (vInv.size() > MAX_INV_SZ || vInv.empty())
         {
+            LOCK(cs_main);
             Misbehaving(pfrom->GetId(), 20);
             return error("message inv size() = %u", vInv.size());
         }
-        
-        for (unsigned int nInv = 0; nInv < vInv.size(); nInv++)  // Validate that INVs are a valid type
+        for (unsigned int nInv = 0; nInv < vInv.size(); nInv++)
         {
             const CInv &inv = vInv[nInv];
-            if (!((inv.type == MSG_TX) || (inv.type == MSG_BLOCK)))
+            if (!((inv.type == MSG_TX) || (inv.type == MSG_BLOCK)) || inv.hash.IsNull())
             {
```

Comment:
Can reduce the nested parens by changing this to `inv.type != MSG_TX && inv.type != MSG_BLOCK` instead.